### PR TITLE
feat(rbac): Audit log rbac backend

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
### What does this pull request do

Implement audit logging for RBAC backend

### Referenced issue:
https://issues.redhat.com/browse/RHIDP-2004

### CODE REVIEW

File: plugins/rbac-backend/src/audit-log/audit-logger.ts [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+export type EvaluationAuditInfo = {
+  userEntityRef: string;
+  permissionName: string;
+  action: PermissionAction;
+  resourceType?: string;
+  decision?: PolicyDecision;
+};
+
+export enum ConditionEvents {
```
Comment (schultzp2020) [inline plugins/rbac-backend/src/audit-log/audit-logger.ts:72]: FYI, enums are considered harmful in TS. https://www.youtube.com/watch?v=jjMbPt_H3RQ

**Review by PatAKnight — COMMENTED** [review]
I am seeing audit logging throughout the plugin (database files, rest api, enforcer, csv file). Would it make more sense to just have the audit log at the highest level (`csv-file-watcher.ts`, `permission-policy.ts`, and `policies-rest-api.ts`).

We could then wrap enforcer calls with a try catch and record successes / failures that way. So something like this, using the `removePermissionPolicies` from `CSVFileWatcher`:

```
  async removePermissionPolicies(): Promise<void> {
    try {
      await this.enforcer.removePolicies(
        this.csvFilePolicies.removedPolicies,
        'csv-file',        
        false,
        true,
      );
      
      const auditOptions = createAuditPermissionOptions(
        this.csvFilePolicies.removedPolicies,
        'DELETE',
        'csv-file',
        'rbac-backend,
        { removed: this.csvFilePolicies.removedPolicies },
      );
       
      await this.auditDefLog.auditLog(auditOptions);
    } catch (error) {
      this.logger.warn(
        `Failed to remove policies ${JSON.stringify(
          this.csvFilePolicies.removedPolicies,
        )} after modification ${this.csvFileName}. Cause: ${e}`,
      );
     
      // Has not been created yet, using this as an example
      const auditOptions = createAuditPermissionErrorOptions(
        this.csvFilePolicies.removedPolicies,
        'DELETE',
        'csv-file',
        'rbac-backend,
        error,
      );
      
      await this.auditDefLog.auditErrorLog(auditOptions);
    }
    this.csvFilePolicies.removedPolicies = [];
  }
```

We could do this for any enforcer call in the `permission-policy.ts`, `policies-rest-api.ts`, and `csv-file-watcher.ts`. As long as we make sure that errors are properly thrown throughout, this should be sufficient for the audit log.

What do you think?

File: plugins/rbac-backend/src/plugin.ts [hunk 2]
```diff
@@ -4,6 +4,7 @@ import {
   createBackendPlugin,
 } from '@backstage/backend-plugin-api';
 
+// import * as util from 'util';
```
Comment (PatAKnight) [inline plugins/rbac-backend/src/plugin.ts:7]: Remove this commented out code

**Comment by AndrienkoAleksandr** [comment]
@PatAKnight I created alternative pr with your proposals: https://github.com/janus-idp/backstage-plugins/pull/1726

**Comment by PatAKnight** [comment]
To fix the failing check, change your title from `feat(rbac): Audit log rbac backend` to `feat(rbac): audit log rbac backend`. Also, you have your commit set as a `fix` instead of a `feat`.

**Comment by AndrienkoAleksandr** [comment]
Closed on favor: https://github.com/janus-idp/backstage-plugins/pull/1726

**Comment by AndrienkoAleksandr** [comment]
schultzp2020, take a look, please https://github.com/janus-idp/backstage-plugins/pull/1726

**Pull request closed without merging**